### PR TITLE
gh-358: add static types support

### LINKED ISSUE #358: Add static types
### Is Your Feature Request Related to a Problem? Please Describe

I've split #280 up into two issues. This issue is for specifically adding static typing - and fixing some `mypy` errors where possible.

### Describe the Solution You'd Like

_No response_

### Describe Alternatives You've Considered

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by paddyroddy**
Closes #358. Replaces #308.

This PR adds static type support, hence satisfying `mypy`. With the adaptation of #67 I anticipate typing to change, but it's much easier to do once `mypy` is passing initially.

The only `# type: ignore[]` comments remaining are `attr-defined` which are all attributed to `np.trapz`, those can easily be fixed by #207.

I have turned off:
- `disallow_untyped_decorators` this is being triggered by `pytest` and `nox`. The equivalent `# type: ignore[]` error is `misc` which isn't the most informative
- `warn_return_any` which is due to external libraries like `cosmology` - this could be revisited in the future

I have decided to re-write some of the tests which were simply not following the typing - like passing is a `list` rather than a `np.ndarray`. The alternative approach would be having really messy typing.

I'm not expecting this to be fully perfect. We will likely have to iterate on this, and completely change once we follow #67. But this is a good first step.

Do make sure to look carefully. Hard to avoid such a big PR with this. There were over 200 `# type: ignore[]` comments to fix!

### CODE REVIEW

**Comment by paddyroddy** [comment]
FYI I'm leaving this non-draft as I accidentally broke the tests

File: .git-blame-ignore-revs
Comment (paddyroddy) [inline .git-blame-ignore-revs:7]: A reminder I should add the hash once merged

File: pyproject.toml
Comment (paddyroddy) [inline pyproject.toml:108]: There are 17 errors which are ignored by this. They are almost all related to untyped external libraries. Do we want this here or the individual `# type: ignore[no-any-return]` comments?

File: tests/test_shells.py
Comment (paddyroddy) [inline tests/test_shells.py:33]: I made the default `zeff` to be `0` as `None` didn't make sense as a type

File: tests/test_fits.py
Comment (paddyroddy) [inline tests/test_fits.py:27]: Move this here as I wanted `fitsio` for typing

File: glass/user.py
Comment (paddyroddy) [inline glass/user.py:34]: This is a bit hacky, but seems to work for `mypy`

File: glass/user.py
Comment (paddyroddy) [inline glass/user.py:81]: Now `self.ext` defaults to `""` then no harm in having a blank extension

File: glass/fields.py
Comment (paddyroddy) [inline glass/fields.py:47]: I got rid of these as I don't really like custom types. Can return if there are strong feelings.

Comment (paddyroddy) [reply glass/fields.py:47]: Although I suspect `mypy` won't pass if all have the same type, i.e. `Cls` do vary depending on context

File: glass/shells.py
Comment (paddyroddy) [inline glass/shells.py:719]: `sum(<x>)` is equivalent to `np.sum(<x>, axis=0)` and is more accurately showing what is happening + fixes `mypy`

File: glass/user.py
Comment (ntessore) [reply glass/user.py:81]: But those are different situations: a blank EXTNAME is not the same as not having an EXTNAME.

File: docs/conf.py
Comment (ntessore) [inline docs/conf.py:1]: Would also make sense to me to include only `glass/` in the type checking.

File: examples/2-advanced/cosmic_shear.ipynb
Comment (ntessore) [inline examples/2-advanced/cosmic_shear.ipynb:1]: Why are the notebooks being formatted in this PR?

File: .gitignore
Comment (ntessore) [inline .gitignore:1]: Not sure this is really a change that needs to happen in this PR.

File: glass/core/array.py
Comment (ntessore) [inline glass/core/array.py:114]: better to use [DTypeLike](https://numpy.org/doc/stable/reference/typing.html#numpy.typing.DTypeLike)

File: glass/fields.py
Comment (ntessore) [inline glass/fields.py:44]: this was correct before: `size` has type `int | tuple[int, ...] | None` and default should remain `None`, which is semantically different from `()`

File: glass/fields.py
Comment (ntessore) [inline glass/fields.py:98]: this was correct before: `cls` are a sequence of arrays or sequence of float, not one array

File: glass/fields.py
Comment (ntessore) [inline glass/fields.py:114]: `None` entries must be supported here for backward compatibility

File: glass/fields.py
Comment (ntessore) [inline glass/fields.py:134]: same as above, these types are incorrect now

File: glass/fields.py
Comment (ntessore) [inline glass/fields.py:141]: `None` should be accepted here for backward compatibility

File: glass/fields.py
Comment (ntessore) [inline glass/fields.py:156]: This should allow arbitrary sequences, not just lists

File: glass/fields.py
Comment (ntessore) [inline glass/fields.py:198]: see above

File: glass/user.py
Comment (paddyroddy) [reply glass/user.py:81]: But the function (as I understand it) will add "", which if it already has one isn't going to do anything. I can change this to

```py
if not self.ext or self.ext not in self.fits:
```

it just felt redundant

File: glass/fields.py
Comment (ntessore) [inline glass/fields.py:206]: see above, I don't understand how these pass currently?

File: glass/fields.py
Comment (ntessore) [inline glass/fields.py:211]: This yields the real-valued maps, not the complex-valued alms

File: glass/fields.py
Comment (ntessore) [inline glass/fields.py:248]: `None` entries should be supported for backward compatibility

File: glass/fields.py
Comment (ntessore) [inline glass/fields.py:291]: see above

File: glass/fields.py
Comment (ntessore) [inline glass/fields.py:297]: see above

File: glass/fields.py
Comment (ntessore) [inline glass/fields.py:320]: see above

File: glass/fields.py
Comment (ntessore) [inline glass/fields.py:355]: see above

File: glass/fields.py
Comment (ntessore) [inline glass/fields.py:413]: I find this way harder to read, but I understand you don't want to manually type `pairs` here

File: glass/galaxies.py
Comment (ntessore) [inline glass/galaxies.py:121]: This change should not be necessary if `broadcast_leading_axes()` is typed correctly

File: glass/lensing.py
Comment (ntessore) [inline glass/lensing.py:375]: this works with any sequence though

File: glass/lensing.py
Comment (ntessore) [inline glass/lensing.py:389]: see above, no need to restrict it to a list (and it's not always a list in practice)

File: glass/observations.py
Comment (ntessore) [inline glass/observations.py:86]: these should accept `float` since they are in practice often called with single float arguments

**Review by ntessore — COMMENTED** [review]
First and foremost: monumental work @paddyroddy!

I only managed to make it down to `observations.py` before I have to jump on the next round of calls, but leaving a first set comments here in the meantime. Main issue I've seen is that a) some functions changed behaviour, and b) the type of `cls`/`gls` arguments is incorrect: these are ragged sequences of arrays, for example `[np.array(...), np.array(...), []]` for the `cls` of two shells with no cross-correlation.

Anyway, **excellent** work and titanic effort! :)

File: glass/user.py
Comment (paddyroddy) [reply glass/user.py:81]: Have changed to `if not self.ext or self.ext not in self.fits:`

File: docs/conf.py
Comment (paddyroddy) [reply docs/conf.py:1]: As discussed, it is only possible to exclude modules. Excluding folders doesn't seem supported https://github.com/python/mypy/issues/17977

File: .gitignore
Comment (paddyroddy) [reply .gitignore:1]: Agreed, but can we leave it? I had to run them to see the outputs for typing

File: examples/2-advanced/cosmic_shear.ipynb
Comment (paddyroddy) [reply examples/2-advanced/cosmic_shear.ipynb:1]: It was related to #362. I briefly commented out those rules because with all the typing, it really made sense to force stuff onto their own line. I can revert the notebooks, but that was why.

File: glass/fields.py
Comment (paddyroddy) [reply glass/fields.py:297]: `mypy` actually told me this one before I got around to this comment!

File: glass/fields.py
Comment (paddyroddy) [reply glass/fields.py:413]: If it helps, they are faster too https://switowski.com/blog/for-loop-vs-list-comprehension/#why-is-list-comprehension-faster-than-a-for-loop

File: glass/galaxies.py
Comment (paddyroddy) [reply glass/galaxies.py:121]: Could we leave this now for now and fix properly in #386?

File: glass/fields.py
Comment (paddyroddy) [reply glass/fields.py:98]: So to clarify the expected type is

```python
collections.abc.Sequence[npt.NDArray[np.float64] | collections.abc.Sequence[float]]
```

not

```python
npt.NDArray[np.float64] | collections.abc.Sequence[collections.abc.Sequence[float]]
```

?

File: glass/fields.py
Comment (paddyroddy) [reply glass/fields.py]: As discussed, I'm deprecating the `None` option. But have restored the type as `int | tuple[int, ...]`

File: glass/fields.py
Comment (paddyroddy) [reply glass/fields.py:147]: As discussed, I'm deprecating the `None` option. But have restored the type as `int | tuple[int, ...]`

File: glass/fields.py
Comment (paddyroddy) [reply glass/fields.py:260]: As discussed, I'm deprecating the `None` option. But have restored the type as `int | tuple[int, ...]`

**Comment by paddyroddy** [comment]
I've addressed your comments @ntessore

File: pyproject.toml
Comment (Saransh-cpp) [inline pyproject.toml:137]: 🎉

File: pyproject.toml
Comment (Saransh-cpp) [reply pyproject.toml:108]: I will suggest adding individual # type: ignore[no-any-return]` comments to ensure that our functions don't violate this.

**Review by Saransh-cpp — COMMENTED** [review]
Amazing work 🤯

File: .git-blame-ignore-revs
Comment (paddyroddy) [reply .git-blame-ignore-revs]: Okay, have removed.

File: pyproject.toml
Comment (paddyroddy) [reply pyproject.toml]: Yeah makes sense, have done this.

File: examples/2-advanced/cosmic_shear.ipynb
Comment (paddyroddy) [reply examples/2-advanced/cosmic_shear.ipynb:1]: These have been addressed in #393 instead

File: glass/shells.py
Comment (ntessore) [inline glass/shells.py:61]: I'm not sure eliminating this type alias makes things more readable -- to me, it makes it harder to understand the signatures below.

File: glass/shells.py
Comment (ntessore) [inline glass/shells.py:531]: What happened here?

File: glass/shells.py
Comment (ntessore) [inline glass/shells.py:590]: Is that ruff's doing? I really dislike this, it wasn't exactly a long line before...

File: glass/shells.py
Comment (ntessore) [inline glass/shells.py:737]: No need to use `np.sum()` and hold all of the intermediate arrays in memory at the same time.

File: glass/user.py
Comment (ntessore) [reply glass/user.py:34]: But what's the point of the conditional import here?

File: glass/user.py
Comment (ntessore) [inline glass/user.py:88]: These are not the same, `ext` can be ""

File: glass/user.py
Comment (ntessore) [inline glass/user.py:90]: Same here

File: glass/user.py
Comment (ntessore) [inline glass/user.py:123]: Not the same as `ext=None`.

File: docs/conf.py
Comment (ntessore) [reply docs/conf.py:1]: Would it not suffice to add `packages = "heracles"` to `[tool.mypy]`?

Comment (paddyroddy) [reply docs/conf.py:1]: Just tried that and it didn't work... 🤔

File: glass/shells.py
Comment (paddyroddy) [reply glass/shells.py:531]: I was trying to make it clear that the `# type: ignore[attr-defined]` is specific to `np.trapz`. This is "fixed" in #391

File: glass/shells.py
Comment (paddyroddy) [reply glass/shells.py:590]: https://github.com/glass-dev/glass/pull/368#discussion_r1831118703

File: glass/user.py
Comment (paddyroddy) [reply glass/user.py:123]: I'm aware by what would `ext` ever be `None`? My understanding is "a.csv" + "" = "a.csv"

File: glass/user.py
Comment (paddyroddy) [reply glass/user.py:90]: https://github.com/glass-dev/glass/pull/368#discussion_r1831121217

File: glass/user.py
Comment (paddyroddy) [reply glass/user.py:88]: https://github.com/glass-dev/glass/pull/368#discussion_r1831121217

File: glass/user.py
Comment (ntessore) [reply glass/user.py:123]: Oh, right. This `ext=` is the FITS extension name. Extensions don't need to have a name, so that's why `None` (no name, use extension number) and `""` (empty string) are different

Comment (paddyroddy) [reply glass/user.py:123]: Okay, makes sense. I couldn't find much documentation on FITS, will change

File: glass/shapes.py
Comment (ntessore) [inline glass/shapes.py:96]: 👀 Is that a stray comma at the end of the line that causes mypy to complain?

Comment (paddyroddy) [reply glass/shapes.py:96]: No. The `no-any-return` I believe is because the result of `np.sqrt` is meant to be `npt.NDArray[np.float64]` but the actual result isn't guaranteed to be one. So we could use `.astype`. This is why I haven't fixed that one.

File: glass/shapes.py
Comment (paddyroddy) [reply glass/shapes.py:34]: Okay fair enough

File: glass/shells.py
Comment (paddyroddy) [reply glass/shells.py:61]: Okay, your call. I'm not a fan of the custom types.

Comment (paddyroddy) [reply glass/shells.py:61]: Although things like `WeightFunc` currently don't all have the same signature. So will require some fiddling.

File: glass/user.py
Comment (paddyroddy) [reply glass/user.py:34]: If I directly import `fitsio` then users _have_ to have `fistio` which you're trying to avoid?

File: glass/shells.py
Comment (paddyroddy) [reply glass/shells.py:721]: Not sure I follow? The problem is:

With `sum()` (not numpy one) the assumed output type is `int` so signature is wrong

The `[]` encapsulating the expression is because `mypy` errors with `Generator expression must be parenthesized`

**Comment by paddyroddy** [comment]
@ntessore I've addressed through your commments

File: glass/shells.py
Comment (ntessore) [reply glass/shells.py:721]: Ah, got it, this is changed for the implicit type

**Review by ntessore — APPROVED** [review]
Fantastic! Well done!

**Review by Saransh-cpp — APPROVED** [review]
Awesome! Thanks for putting so much efforts into this, @paddyroddy!

**Comment by paddyroddy** [comment]
Thanks both for the review. Will merge this and update the downstream PRs.

**Pull request merged**
